### PR TITLE
feat: detect camunda.secrets references in input mappings at deploy time

### LINKED ISSUE #56557: [Secret Resolution] Secret reference format + parse-time detection
### Description

Define and parse the `camunda.secrets.<name>` reference format and detect references inside FEEL **input-mapping** expressions at deployment; store `{targetPath -> [references]}` on the executable element. References are honored only in input mappings — anywhere else they are treated as string literals (secret-injection prevention). Multiple references per expression are supported.

### Motivation

Deployment-time detection lets the engine know cheaply and deterministically which elements need resolution at activation.

### Acceptance criteria

- [ ] Shared reference parser + pattern
- [ ] References (with target path) stored on the executable model
- [ ] Multiple references per expression supported
- [ ] References outside input mappings stay literal
- [ ] Unit tests

### Breakdown

- Add a shared `SecretReference` parser (pattern `camunda\.secrets\.<name>`).
- Hook `VariableMappingTransformer#transformInputMappings` + `FlowNodeTransformer#transformIoMappings` (`zeebe/engine`) to scan the raw input-mapping source.
- Add a `secretReferences` field (+ getter/setter) to `ExecutableFlowNode`; persist in the parsed model.
- Unit tests: single/multiple refs; ref used outside an input mapping stays literal.

### PR DESCRIPTION
**Pull request by berkaycanbc**
## What

First implementation task of Centralized Secret Resolution: detect
`camunda.secrets.<name>` references in BPMN **input mappings** at deployment time and store them
on the executable model, so a later task can resolve them at job activation. Detection only — no
resolution, cache, or gateway API in this PR.

## Why these decisions

- **Only input mappings, only expressions.** References are recognized only in input-mapping
  sources that are FEEL expressions (start with `=`). A static value, or a reference written inside
  a FEEL string literal (`="camunda.secrets.x"`), stays literal. This prevents secret injection: a
  runtime value that merely looks like a reference cannot trigger resolution.
- **Single literal-aware regex.** The parser uses one pattern that matches either a whole string
  literal or a real reference; only the reference branch is captured, so references inside quotes
  are consumed and ignored — no manual string scanning.
- **Typed `SecretReference` + `Set`.** Each reference is a typed value; a `Set` de-duplicates
  repeated references within one expression.
- **JSON pointer keys.** The map is keyed by the mapping target as a JSON pointer
  (`tokens.token` → `/tokens/token`, RFC 6901), not dot notation. Job variables are a Jackson
  document, so at activation the value can be replaced natively via `JsonPointer` with zero path
  conversion on the hot path; the dot→slash conversion happens once, at deploy time.

## Commits

1. `feat: add secret reference parser and model field`
2. `feat: detect secret references in input mappings during transformation`
3. `test: cover secret reference parsing and detection`

## Testing

Added unit tests for the parser and the transformation: single/multiple references and
de-duplication, string-literal and static-value exclusion, input-vs-output-mapping scoping, and
JSON-pointer keys including nested target paths.

Closes #56557

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+   *   <li>{@code =xcamunda.secrets.token} → nothing (part of a longer identifier)
+   *   <li>{@code =order.camunda.secrets.token} → nothing (nested property, not the root)
+   * </ul>
+   */
+  private static final Pattern REFERENCE_PATTERN =
+      Pattern.compile(
+          "\"(?:\\\\.|[^\"\\\\])*\""
+              + "|(?<![\\w.])"
+              + Pattern.quote(PREFIX)
+              + "([a-zA-Z_][a-zA-Z0-9_]*)");
```
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: 😵 That's one hell of a regex. I hope I find a lot of tests for this in the next commits 😄

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    final var references = new LinkedHashSet<SecretReference>();
+    final Matcher matcher = REFERENCE_PATTERN.matcher(source);
+    while (matcher.find()) {
+      // group 1 is null when the match is a string literal, which must be ignored
+      final var name = matcher.group(1);
+      if (name != null) {
+        references.add(new SecretReference(name));
+      }
```
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: 🔧 Can we pull this null check outside of the while loop? No need to check group 1 for every referenced secret, right?

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeTransformer.java [hunk 3]
```diff
@@ -86,6 +86,14 @@ private void transformIoMappings(
...
                 variableMappingTransformer.transformOutputMappings(mappings, expressionLanguage))
         .ifPresent(flowNode::setOutputMappings);
+
+    // secret references are only honored in input mappings; anywhere else they stay literal
+    ioMapping
+        .map(ZeebeIoMapping::getInputs)
+        .filter(mappings -> !mappings.isEmpty())
+        .map(variableMappingTransformer::detectSecretReferences)
+        .filter(references -> !references.isEmpty())
+        .ifPresent(flowNode::setSecretReferences);
```
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeTransformer.java]: 🙃 Would be nice to group this together with the other input mapping section. But extremely nitpicky so take it or leave it.

Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeTransformer.java]: 🤔 I wonder if we could just set it on the input mappings itself and detect the secret references during the `variableMappingTransformer.transformInputMappings(mappings, expressionLanguage)` step.

I think it'd be cleaner as you know exactly which input mapping contains which secret references. The input mapping also already contains the target.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java [hunk 4]
```diff
@@ -95,6 +97,41 @@ public Expression transformOutputMappings(
...
+   * becomes a pointer segment; {@code ~} and {@code /} inside a segment are escaped per RFC 6901
+   * ({@code ~} -> {@code ~0}, {@code /} -> {@code ~1}).
+   */
+  private static String toJsonPointer(final String targetPath) {
+    final var pointer = new StringBuilder();
+    for (final String segment : targetPath.split("\\.")) {
+      pointer.append('/').append(segment.replace("~", "~0").replace("/", "~1"));
+    }
+    return pointer.toString();
+  }
```
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java:191]: ❓ What is this and what does it do? I don't understand it.

Comment (remcowesterhoud) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java:191]: Copilot told me this is so we can do a targeted json lookup later. Can we document this more clearly in the javadoc?

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java [hunk 5]
```diff
@@ -95,6 +97,41 @@ public Expression transformOutputMappings(
...
+   * (non-expression) values are ignored (see {@link SecretReference}). Mappings without any
+   * reference are omitted from the result.
+   */
+  public Map<String, Set<SecretReference>> detectSecretReferences(
+      final Collection<? extends ZeebeMapping> inputMappings) {
+    final var secretReferences = new LinkedHashMap<String, Set<SecretReference>>();
+    for (final ZeebeMapping mapping : inputMappings) {
+      final var references = SecretReference.parse(mapping.getSource());
+      if (!references.isEmpty()) {
+        secretReferences.put(toJsonPointer(mapping.getTarget()), references);
```
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java]: ❌ I think the target is to broad, isn't it? If I have input mapping:

`foo -> {x1: "camunda.secrets.x", x2: camunda.secrets.x}`

Don't we want the json pointer for the secret to be `foo/x2`?

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    return Stream.of(
+        arguments("=camunda.secrets.token", refs("token")),
+        arguments("=\"Bearer \" + camunda.secrets.token", refs("token")),
+        arguments(
+            "=\"Bearer \" + camunda.secrets.token + camunda.secrets.postfix",
+            refs("token", "postfix")),
+        arguments("=camunda.secrets.MY_SECRET_2", refs("MY_SECRET_2")),
+        arguments("=camunda.secrets._underscore", refs("_underscore")),
+        // a literal reference is ignored, but the expression reference in the same source is parsed
+        arguments("=\"camunda.secrets.literal\" + camunda.secrets.real", refs("real")),
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java:186]: 🔧 Let's have a case with just the literal if that's possible.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        // reference nested inside a FEEL context value
+        arguments("={ token: camunda.secrets.token }", refs("token")),
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java:188]: 🔧 Can we get some more test cases with these FEEL contexts? I'm thinking a bigger context with multiple references in there. Maybe even a mix of literals an non-literals.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    // when
+    final var references = SecretReference.parse(source);
+
+    // then
+    assertThat(references).isEqualTo(expected);
+  }
+
+  @ParameterizedTest(name = "[{index}] {0}")
+  @NullSource
+  @ValueSource(
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java:45]: 🔧 Let's add FEEL context cases here as well.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+                t.zeebeInputExpression(
+                    "\"Bearer \" + camunda.secrets.token", "tokens.externalSystemToken"));
+
+    // when
+    final var secretReferences = task.getSecretReferences();
+
+    // then - the dotted target path is stored as a JSON pointer
+    assertThat(secretReferences)
+        .containsExactly(
+            entry("/tokens/externalSystemToken", Set.of(new SecretReference("token"))));
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java:54]: ❓ Can you remind, will we just do a replace of the secret reference with the resolved secret at this path? I'm wondering if we need the `Bearer ` information in our secret reference as well, but I don't think we do.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+import io.camunda.zeebe.model.bpmn.BpmnModelInstance;
+import io.camunda.zeebe.model.bpmn.builder.ServiceTaskBuilder;
+import java.time.InstantSource;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import org.junit.jupiter.api.Test;
+
+class FlowNodeSecretReferenceTest {
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java:30]: 🔧 Probably a stupid case, but can we add a case where `camunda.secrets.x` is the target instead of the source.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+ * <p>A single source may contain multiple references; each distinct reference is returned once.
+ */
+@NullMarked
+public record SecretReference(String name) {
+
+  /** The FEEL expression marker; a source is only an expression when it starts with this. */
+  private static final String EXPRESSION_MARKER = "=";
+
+  /** The prefix that identifies a secret reference. */
+  private static final String PREFIX = "camunda.secrets.";
```
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: 💭  Maybe a bit stupid, but in FEEL it's also valid to add whitespace and write for example"

`apiKey -> secrets . camunda . apiKey`

If this is done we won't list a secret reference. This means we also won't replace it so we don't leak anything. But I wanted to highlight this limitation we have.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java (hunk 1 above)
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: ❌ FEEL supports more characters than just these. With this if I have a secret called: `camunda.secrets.tokén` my secret reference becomes `camunda.secrets.tok`. Please add a test case for special characters as well.

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
@berkaycanbc I had Fable look for some gaps in testing and edge-cases regarding the regex. This is what it came up with it.

It has some decent points. Let's make sure we have a loooot of test coverage here. If we get false positives we can be leaking secrets.

**Personally I think mainly supporting more characters in the secret name and test gaps are important.**

Ignore the `FEEL resolves the secret` part. That's not true because FEEL won't be resolving the secrets.

--------

Verified everything empirically against feel-scala 1.21.0 (the exact version this repo ships) by running the real FEEL engine next to a copy of the
  regex. There are genuine unknown unknowns — the tests all pass, but they encode assumptions about FEEL syntax that the actual parser doesn't share.
  Here's what I found, ordered by how much it matters.

  ### **Mismatches between FEEL and the regex (verified, not speculation)**

  1. **Whitespace/newlines around the dots** — false negative. FEEL tokenizes path expressions, so whitespace around . is insignificant:

```
  = camunda . secrets . token      → FEEL resolves the secret, parse() returns []
  = camunda\n  .secrets\n  .token  → FEEL resolves the secret, parse() returns []
```

  This is the most realistic miss: anyone writing a longer multi-line input mapping (or using a formatter that breaks before .) gets a reference that
  evaluates as a secret but is invisible to detection. The regex requires the contiguous literal camunda.secrets..

  2. **Unicode identifiers** — wrong name captured, worse than a miss. FEEL names follow the DMN name-char spec, which includes unicode letters:

```
  =camunda.secrets.tokén   → FEEL resolves secret "tokén", parse() returns [tok]
```

  The [a-zA-Z0-9_]* name class stops at é, so you don't get nothing — you get a truncated, wrong secret name. Depending on what the consumer does with
  these (fetch? authorize?), that's fetching the wrong secret and missing the right one simultaneously.

  3. **Backticked names** — false negative. FEEL's escape hatch for names with special characters works on the path segment:

```
  =camunda.secrets.`my-secret`   → FEEL resolves secret "my-secret", parse() returns []
```

  This matters if secret names in the store can contain characters like - (connector secret conventions allow it) — backticks are then the only way to
  reference such a secret in FEEL, and that entire class is undetectable.

  4. **FEEL comments** — false positive. feel-scala supports // and /* */:

```
  =camunda.secrets.token // camunda.secrets.other   → FEEL uses only token, parse() returns [token, other]
```

  A commented-out reference gets detected as used. Whether that's harmful depends on the consumer — over-fetching/over-authorizing a secret the
  expression never reads, or a task failing an authz check because of a comment.

  5. **Semantically equivalent forms** — false negatives, arguably by design:

```
  =(camunda).secrets.token              → resolves, parse() returns []
  =get value(camunda.secrets, "token")  → resolves, parse() returns []
  =for camunda in [...] return camunda.secrets.token  → loop var shadows the root; parse() returns [token] (false positive)
```

  These are fine to declare out of scope, but the current Javadoc doesn't say so — the contract reads as "references used as expressions are found",
  which items 1–3 already break.

  ### **Plain test gaps** (behavior is correct today, just unpinned)

  - String literal ending in an escaped backslash: ="a\\" + camunda.secrets.x. The regex handles this correctly ((?:\\.|[^"\\])* pairs \\ properly),
  but it's the classic escape-handling trap and the one escaped-quote test doesn't cover it. Cheap to pin.
  - Unterminated string literal: ="camunda.secrets.token — the literal alternative can't match, so the reference is captured from inside the broken
  quote. Probably unreachable because expression validation rejects the deployment before transformation, but worth a comment or test documenting that
  assumption.
  - Trailing path after the name: =camunda.secrets.token.length captures token — correct, untested.
  - Multi-line expression with a reference (the happy case, not the whitespace-path case) — untested.

  The underlying issue

  The regex is re-implementing a subset of the FEEL grammar and will keep drifting from it — every finding above is FEEL syntax the regex doesn't model
  (token whitespace, comments, backticks, unicode name chars). Since detection presumably feeds secret resolution/authorization, a false negative
  isn't just cosmetic. The more robust option: parse the expression with the actual FEEL parser (the engine already parses every mapping expression at
  deployment via zeebe/expression-language) and walk the AST for camunda.secrets.<name> path expressions. That gets 1–4 right for free; string literals
  and comments fall out naturally. Shadowing (item 5) remains a corner case either way.

  If the regex stays, I'd at minimum fix the whitespace and unicode cases, decide explicitly on backticks and comments, and pin each decision with a
  test — plus document the known-undetectable forms in the Javadoc.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java (hunk 1 above)
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: Replaced with FEEL parser.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java (hunk 2 above)
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: Replaced with FEEL parser.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java (hunk 4 above)
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java:191]: updated

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java (hunk 5 above)
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java]: implemented

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java (hunk 6 above)
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java:186]: updated

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java (hunk 7 above)
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java:188]: added more tests

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java (hunk 9 above)
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java:54]: No, we will do a replacement for matched secret refs.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java (hunk 11 above)
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: Replaced with FEEL parser, so should be resolved.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java (hunk 1 above)
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: Replaced with FEEL parser, so should be resolved.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java (hunk 8 above)
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java:45]: updated.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java (hunk 10 above)
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java:30]: added in `shouldNotStoreSecretReferenceUsedAsTarget` test

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeTransformer.java (hunk 3 above)
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeTransformer.java]: Updated, detection now sits directly after the input-mapping transform, before outputs.

Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeTransformer.java]: Implemented in 1c0c206bb9273e5a7d9917a9add637e7d023913e.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java [hunk 12]
```diff
@@ -95,6 +104,57 @@ public Expression transformOutputMappings(
...
+    for (final Mapping mapping : mappings) {
+      for (final var detected : SecretReference.parse(mapping.source)) {
+        secretsByPointer
+            .computeIfAbsent(
+                toJsonPointer(mapping.target, detected.path()), key -> new LinkedHashSet<>())
+            .add(detected.secret());
+      }
+    }
+    return secretsByPointer;
+  }
```
[the remark this answers is not in the document]
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java:172]: Makes sense, updated the code to support that edge case.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java [hunk 13]
```diff
@@ -95,6 +104,57 @@ public Expression transformOutputMappings(
...
+   * Detects the secrets referenced by the already-parsed input mappings, keyed by the JSON pointer
+   * (RFC 6901) of the leaf each secret belongs to — the mapping's dotted target plus the
+   * reference's FEEL context path. Examples: {@code "Bearer " + camunda.secrets.token -> tokens.t}
+   * gives {@code /tokens/t}; {@code {x2: camunda.secrets.x} -> foo} gives {@code /foo/x2}. The
+   * pointer lets job activation replace the reference in the job variables via a Jackson {@code
+   * JsonPointer}; the conversion happens once, here at deploy time.
+   *
+   * <p>Reuses the sources parsed by {@link #toMappings}, so each source is parsed once. Mappings
+   * with no reference are omitted (see {@link SecretReference} for what counts).
+   */
```
[the remark this answers is not in the document]
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java:123]: @remcowesterhoud FYI, this is a limitation that I do not expect we support in this iteration.

Given
```
Input mapping
  source:  = if useBackup
             then { token: camunda.secrets.backupToken }
             else { token: camunda.secrets.primaryToken }
  target:  auth
```

Then the path will be `/auth` instead of `/auth/token`.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java [hunk 14]
```diff
@@ -7,97 +7,127 @@
...
   }
+
+  /**
+   * A secret reference together with the context path where it occurs within a mapping source. The
+   * path holds the keys of any enclosing FEEL context, so it is empty for a scalar source (e.g.
+   * {@code ="Bearer " + camunda.secrets.token}) and {@code [x2]} for a reference nested in a
+   * context (e.g. {@code ={x2: camunda.secrets.token}}). It is later appended to the mapping target
+   * to form the JSON pointer of the leaf the secret value belongs to.
+   */
+  public record Located(List<String> path, SecretReference reference) {}
```
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: 🤔 Not sure about the naming of this one. `Located` seems very indescriptive.

Claude suggested `Occurence` for me, but it's not much better either. So feel free to leave it.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java [hunk 15]
```diff
@@ -7,97 +7,127 @@
...
   /** The full reference as it appears in an expression, e.g. {@code camunda.secrets.token}. */
   public String reference() {
-    return PREFIX + name;
+    return ROOT + "." + NAMESPACE + "." + name;
   }
+
+  /**
+   * A secret reference together with the context path where it occurs within a mapping source. The
+   * path holds the keys of any enclosing FEEL context, so it is empty for a scalar source (e.g.
+   * {@code ="Bearer " + camunda.secrets.token}) and {@code [x2]} for a reference nested in a
```
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: ❓ I suppose we don't need it for a scalar source, because it's stored on the input mapping and we have the target in there already?

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java [hunk 16]
```diff
@@ -103,14 +165,34 @@ static Stream<Arguments> sourcesWithReferences() {
...
         arguments("=camunda.secrets._underscore", refs("_underscore")),
+        // whitespace and line breaks around the dots are insignificant in FEEL
+        arguments("=camunda . secrets . token", refs("token")),
+        arguments("=camunda\n  .secrets\n  .token", refs("token")),
+        // unicode names are valid FEEL identifiers
+        arguments("=camunda.secrets.tokén", refs("tokén")),
+        // backtick-escaped names allow special characters
+        arguments("=camunda.secrets.`my-secret`", refs("my-secret")),
+        // trailing path access after the secret name still references the secret
+        arguments("=camunda.secrets.token.length", refs("token")),
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java]: ❓ How is a case like this supposed to work? I don't think it'll work 😄, but I suppose the ref is still correct.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java [hunk 17]
```diff
@@ -95,6 +95,35 @@ void shouldStoreReferencesFromMultipleInputMappings() {
...
+            t ->
+                t.zeebeInputExpression(
+                    "{x1: \"camunda.secrets.x\", x2: camunda.secrets.x}", "foo"));
+
+    // when
+    final var secretReferences = task.getSecretReferences();
+
+    // then - the pointer targets the entry that references the secret, not the whole target
+    assertThat(secretReferences)
+        .containsExactly(entry("/foo/x2", Set.of(new SecretReference("x"))));
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java:131]: ❤️ Love it!

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java:123]: I think that's fine. We'll still resolve the secret at the end

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java [hunk 18]
```diff
@@ -188,6 +188,56 @@ void shouldHaveEmptySecretReferencesWithoutInputMappings() {
...
+            t ->
+                t.zeebeInputExpression("camunda.secrets.s1", "a")
+                    .zeebeInputExpression("camunda.secrets.s2", "a.b"));
+
+    // when
+    final var secretReferences = task.getSecretReferences();
+
+    // then - the overridden scalar 'a' contributes no secret; only the surviving /a/b remains
+    assertThat(secretReferences).containsExactly(entry("/a/b", Set.of(new SecretReference("s2"))));
+  }
```
Comment (remcowesterhoud) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java:242]: 🔧 Could we add a test like:

```java
    final var task =
        transform(
            t ->
                t.zeebeInputExpression("camunda.secrets.s1", "a.b")
                    .zeebeInputExpression("camunda.secrets.s2", "a.c"));

    // when
    final var secretReferences = task.getSecretReferences();

    assertThat(secretReferences)
        .containsExactly(
            entry("/a/b", Set.of(new SecretReference("s1"))),
            entry("/a/c", Set.of(new SecretReference("s2"))));
```

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java [hunk 19]
```diff
@@ -95,6 +104,68 @@ public Expression transformOutputMappings(
...
+      final var targetPath = new ArrayList<>(context.path);
+      targetPath.add(entry.getKey());
+      for (final var detected : SecretReference.parse((Expression) entry.getValue())) {
+        final var leafPath = new ArrayList<>(targetPath);
+        leafPath.addAll(detected.path());
+        secretsByPointer
+            .computeIfAbsent(toJsonPointer(leafPath), key -> new LinkedHashSet<>())
+            .add(detected.secret());
+      }
+    }
```
Comment (remcowesterhoud) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java]: 🔧 We could ulitize the `MappingContext.visit` method here. That way the looping is taken away from your logic here.

**Review by remcowesterhoud — APPROVED** [review]
@berkaycanbc I'll pre-approve to not lbock further. I have a few comments, please look at those before merging 🙂

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java (hunk 14 above)
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: @remcowesterhoud yeah, I didn't like `Located` as well and renamed it to `DetectedSecret` to be more explicit.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java (hunk 15 above)
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: @remcowesterhoud It is actually empty at the moment that `DetectedSecret` is created. But I find it confusing as well, I'll remove it.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java (hunk 16 above)
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java]: I'll check if this is something that we can prevent easily during parsing because the feel evaluation will not convert this to a literal string but try to evaluate it.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java (hunk 14 above)
Comment (remcowesterhoud) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReference.java]: Sorry, I didn't remove the comment. I noticed it was outdated in a later commint :D

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java (hunk 16 above)
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/element/SecretReferenceTest.java]: Resolved in a7140b81eedfe9993f61ef81140e04f0a5582110

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java (hunk 18 above)
Comment (berkaycanbc) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/FlowNodeSecretReferenceTest.java:242]: resolved in ac9e4974c9b85993d2df0e0ee6fea82f1bdec092

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java (hunk 19 above)
Comment (berkaycanbc) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/transformer/VariableMappingTransformer.java]: resolved in f870f6be63151a0fd21e721bf305507f348e7b76
